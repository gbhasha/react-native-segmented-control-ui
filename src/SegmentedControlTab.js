import React from 'react'
import { View, ViewPropTypes, StyleSheet,  Text } from 'react-native'
import PropTypes from 'prop-types'
import TabOption from './TabOption'

export const handleTabPress = (index, selectedIndex, onTabPress) => {
    selectedIndex !== index &&  onTabPress(index)
}

export const getAccessibilityLabelByIndex = (accessibilityLabels, index) => {
    return accessibilityLabels && accessibilityLabels.length > 0 && accessibilityLabels[index] ?  accessibilityLabels[index] : undefined
}

const SegmentedControlTab = ({
    selectedIndex, values,
    badges, borderRadius, tabsContainerStyle,
    tabStyle, activeTabStyle,
    tabTextStyle, activeTabTextStyle,
    tabBadgeContainerStyle, activeTabBadgeContainerStyle,
    tabBadgeStyle, activeTabBadgeStyle,
    onTabPress, textNumberOfLines,
    allowFontScaling,
    accessible,
    accessibilityLabels,
    testIDs,
}) => {

    const firstTabStyle = [{ borderLeftWidth: 0, borderTopWidth: 0}]
    const lastTabStyle = [{ }]
    const containerStyle = [{borderRadius: borderRadius, overflow: 'hidden'}]
    return (
        <View
            style={[styles.tabsContainerStyle, containerStyle, tabsContainerStyle]}
            removeClippedSubviews={false}>
            {
                values.map((item, index) => {
                    const accessibilityText = getAccessibilityLabelByIndex(accessibilityLabels, index)
                    return (
                        <TabOption
                            key={index}
                            index={index}
                            testID={testIDs.length > 0 ? testIDs[index] : undefined}
                            badge={badges && badges[index] ? badges[index] : undefined}
                            isTabActive={selectedIndex === index}
                            text={item}
                            textNumberOfLines={textNumberOfLines}
                            onTabPress={(index) => handleTabPress(index, selectedIndex, onTabPress)}
                            firstTabStyle={index === 0 ? [firstTabStyle] : {}}
                            lastTabStyle={index === values.length - 1 ? [lastTabStyle] : {}}
                            tabStyle={[tabStyle]}
                            activeTabStyle={activeTabStyle}
                            tabTextStyle={tabTextStyle}
                            activeTabTextStyle={activeTabTextStyle}
                            tabBadgeContainerStyle={tabBadgeContainerStyle}
                            activeTabBadgeContainerStyle={activeTabBadgeContainerStyle}
                            tabBadgeStyle={tabBadgeStyle}
                            activeTabBadgeStyle={activeTabBadgeStyle}
                            allowFontScaling={allowFontScaling}
                            accessible={accessible}
                            accessibilityLabel={accessibilityText ? accessibilityText : item }
                        />
                    )
                })
            }
        </View>
    )
}

SegmentedControlTab.propTypes = {
    accessibilityLabels: PropTypes.arrayOf(PropTypes.string),
    accessible: PropTypes.bool,
    activeTabBadgeContainerStyle: Text.propTypes.style,
    activeTabBadgeStyle: Text.propTypes.style,
    activeTabStyle: ViewPropTypes.style,
    activeTabTextStyle: Text.propTypes.style,
    allowFontScaling: PropTypes.bool,
    badges: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ])),
    borderRadius: PropTypes.number,
    onTabPress: PropTypes.func,
    selectedIndex: PropTypes.number,
    tabBadgeContainerStyle: Text.propTypes.style,
    tabBadgeStyle: Text.propTypes.style,
    tabStyle: ViewPropTypes.style,
    tabTextStyle: Text.propTypes.style,
    tabsContainerStyle: ViewPropTypes.style,
    testIDs: PropTypes.arrayOf(PropTypes.string),
    textNumberOfLines: PropTypes.number,
    values: PropTypes.arrayOf(PropTypes.string),
}

SegmentedControlTab.defaultProps = {
    values: ['One', 'Two', 'Three'],
    accessible: true,
    accessibilityLabels: [],
    testIDs: [],
    badges: [],
    selectedIndex: 0,
    onTabPress: () => { },
    tabsContainerStyle: {},
    tabStyle: {},
    activeTabStyle: {},
    tabTextStyle: {},
    activeTabTextStyle: {},
    tabBadgeContainerStyle: {},
    activeTabBadgeContainerStyle: {},
    tabBadgeStyle: {},
    activeTabBadgeStyle: {},
    borderRadius: 5,
    textNumberOfLines: 1,
    allowFontScaling: true,
}

const styles = StyleSheet.create({
    tabsContainerStyle: {
        backgroundColor: 'transparent',
        flexDirection: 'row',
        borderColor: '#007AFF',
        borderWidth: 1,
    },
})

export default SegmentedControlTab
