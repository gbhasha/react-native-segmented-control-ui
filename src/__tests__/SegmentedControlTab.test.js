// __tests__/Intro-test.js
import React from 'react';
import SegmentedControlTab from '../SegmentedControlTab';
jest.mock("../TabOption", () => 'TabOption')

import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<SegmentedControlTab />).toJSON();
  expect(tree).toMatchSnapshot();
});