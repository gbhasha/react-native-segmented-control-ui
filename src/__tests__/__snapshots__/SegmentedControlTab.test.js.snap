// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  removeClippedSubviews={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderColor": "#007AFF",
        "borderWidth": 1,
        "flexDirection": "row",
      },
      Array [
        Object {
          "borderRadius": 5,
          "overflow": "hidden",
        },
      ],
      Object {},
    ]
  }
>
  <TabOption
    accessibilityLabel="One"
    accessible={true}
    activeTabBadgeContainerStyle={Object {}}
    activeTabBadgeStyle={Object {}}
    activeTabStyle={Object {}}
    activeTabTextStyle={Object {}}
    allowFontScaling={true}
    firstTabStyle={
      Array [
        Array [
          Object {
            "borderLeftWidth": 0,
            "borderTopWidth": 0,
          },
        ],
      ]
    }
    index={0}
    isTabActive={true}
    lastTabStyle={Object {}}
    onTabPress={[Function]}
    tabBadgeContainerStyle={Object {}}
    tabBadgeStyle={Object {}}
    tabStyle={
      Array [
        Object {},
      ]
    }
    tabTextStyle={Object {}}
    text="One"
    textNumberOfLines={1}
  />
  <TabOption
    accessibilityLabel="Two"
    accessible={true}
    activeTabBadgeContainerStyle={Object {}}
    activeTabBadgeStyle={Object {}}
    activeTabStyle={Object {}}
    activeTabTextStyle={Object {}}
    allowFontScaling={true}
    firstTabStyle={Object {}}
    index={1}
    isTabActive={false}
    lastTabStyle={Object {}}
    onTabPress={[Function]}
    tabBadgeContainerStyle={Object {}}
    tabBadgeStyle={Object {}}
    tabStyle={
      Array [
        Object {},
      ]
    }
    tabTextStyle={Object {}}
    text="Two"
    textNumberOfLines={1}
  />
  <TabOption
    accessibilityLabel="Three"
    accessible={true}
    activeTabBadgeContainerStyle={Object {}}
    activeTabBadgeStyle={Object {}}
    activeTabStyle={Object {}}
    activeTabTextStyle={Object {}}
    allowFontScaling={true}
    firstTabStyle={Object {}}
    index={2}
    isTabActive={false}
    lastTabStyle={
      Array [
        Array [
          Object {},
        ],
      ]
    }
    onTabPress={[Function]}
    tabBadgeContainerStyle={Object {}}
    tabBadgeStyle={Object {}}
    tabStyle={
      Array [
        Object {},
      ]
    }
    tabTextStyle={Object {}}
    text="Three"
    textNumberOfLines={1}
  />
</View>
`;
